pkg/istio-agent/sds-agent.go

```diff
@@ -0,0 +1,204 @@
+func StartSDS() (*sds.Server, error) {
+	applyEnvVars()
+	gatewaySdsCacheOptions = workloadSdsCacheOptions
+
+	// Next to the envoy config, writeable dir (mounted as mem)
+	serverOptions.WorkloadUDSPath = "./var/lib/istio/proxy/SDS"
+	// TODO: remove the caching, workload has a single cert
+	workloadSecretCache := newSecretCache(serverOptions)
+
+	var gatewaySecretCache *cache.SecretCache
+	if serverOptions.EnableIngressGatewaySDS {
+		 gatewaySecretCache = newSecretCache(serverOptions)
+	}
+
+	server, err := sds.NewServer(serverOptions, workloadSecretCache, gatewaySecretCache)
+	if err != nil {
+		return nil, err
+	}
+
+	return server, nil
+}
+
+// newSecretCache creates the cache for workload secrets and/or gateway secrets.
+func newSecretCache(serverOptions sds.Options) (workloadSecretCache *cache.SecretCache) {
+	ret := &secretfetcher.SecretFetcher{}
+
+	// TODO: get the MC public keys from pilot.
+	// TODO: root cert for Istiod from the K8S file or local override
+	// In node agent, a controller is used getting 'istio-security.istio-system' config map
+	// Single caTLSRootCert inside.
+
+	var err error
+	var caClient caClientInterface.Client
+	if "GoogleCA" == serverOptions.CAProviderName {
+		caClient, err = gca.NewGoogleCAClient(serverOptions.CAEndpoint, true)
+	} else {
+		caClient, err = citadel.NewCitadelClient(serverOptions.CAEndpoint, false, nil) // true, rootCert)
+	}
+
+	if err != nil {
+		log.Errorf("failed to create secretFetcher for workload proxy: %v", err)
+		os.Exit(1)
+	}
+	ret.UseCaClient = true
+	ret.CaClient = caClient
+
+		workloadSdsCacheOptions.TrustDomain = serverOptions.TrustDomain
+		workloadSdsCacheOptions.Plugins = sds.NewPlugins(serverOptions.PluginNames)
+		workloadSecretCache = cache.NewSecretCache(ret, sds.NotifyProxy, workloadSdsCacheOptions)
+
+		return workloadSecretCache
+}
+
+// TODO: use existing 'sidecar/router' config to enable loading Secrets
+func newIngressSecretCache(serverOptions sds.Options) (gatewaySecretCache *cache.SecretCache) {
+	gSecretFetcher := &secretfetcher.SecretFetcher{}
+
+	gSecretFetcher.UseCaClient = false
+	cs, err := kube.CreateClientset("", "")
+
+
+	if err != nil {
+			log.Errorf("failed to create secretFetcher for gateway proxy: %v", err)
+			os.Exit(1)
+		}
+	gSecretFetcher.FallbackSecretName = "gateway-fallback"
+	gSecretFetcher.InitWithKubeClient(cs.CoreV1())
+
+	gatewaySecretChan = make(chan struct{})
+		gSecretFetcher.Run(gatewaySecretChan)
+		gatewaySecretCache = cache.NewSecretCache(gSecretFetcher, sds.NotifyProxy, gatewaySdsCacheOptions)
+	return gatewaySecretCache
+}
+
+func applyEnvVars() {
+	serverOptions.PluginNames = strings.Split(pluginNamesEnv, ",")
+
+	serverOptions.EnableWorkloadSDS = true
+
+	serverOptions.EnableIngressGatewaySDS = enableIngressGatewaySDSEnv
+	serverOptions.CAProviderName = caProviderEnv
+	serverOptions.CAEndpoint = caEndpointEnv
+	serverOptions.TrustDomain = trustDomainEnv
```

Comment:
Can we start using trust domain value from the MeshConfig?

Reply:
Not here - agent has no direct access to mesh config. Injector has access to mesh config, and creates the env variables in the agent.